File: json_compression.py
import subprocess
import os


def compress_json_with_jq(input_path):
    # 检查输入文件是否存在
    if not os.path.exists(input_path):
        print(f"错误: 文件 {input_path} 不存在。")
        return
    if os.path.isfile(input_path) and input_path.lower().endswith('.json'):
        try:
            # 构建 jq 命令，将结果输出到临时文件
            temp_file = input_path + ".tmp"
            command = f'jq -c "." "{input_path}" > "{temp_file}"'
            subprocess.run(command, shell=True, check=True)

            # 覆盖原文件
            os.replace(temp_file, input_path)
            print(f"JSON 文件 {input_path} 已成功压缩并覆盖原文件。")
        except subprocess.CalledProcessError as e:
            print(f"执行 jq 命令时出错: {e}")
            # 删除临时文件
            if os.path.exists(temp_file):
                os.remove(temp_file)
        except Exception as e:
            print(f"发生未知错误: {e}")
    elif os.path.isdir(input_path):
        # 处理文件夹中的所有 JSON 文件
        for root, dirs, files in os.walk(input_path):
            for file in files:
                if file.lower().endswith('.json'):
                    file_path = os.path.join(root, file)
                    try:
                        # 构建 jq 命令，将结果输出到临时文件
                        temp_file = file_path + ".tmp"
                        command = f'jq -c "." "{file_path}" > "{temp_file}"'
                        subprocess.run(command, shell=True, check=True)

                        # 覆盖原文件
                        os.replace(temp_file, file_path)
                        print(f"JSON 文件 {file_path} 已成功压缩并覆盖原文件。")
                    except subprocess.CalledProcessError as e:
                        print(f"执行 jq 命令时出错，文件: {file_path}, 错误信息: {e}")
                        # 删除临时文件
                        if os.path.exists(temp_file):
                            os.remove(temp_file)
                    except Exception as e:
                        print(f"发生未知错误，文件: {file_path}, 错误信息: {e}")
    else:
        print(f"错误: {input_path} 不是有效的 JSON 文件或文件夹。")


if __name__ == "__main__":
    input_path = input("请输入要压缩的 JSON 文件或文件夹的路径: ")
    compress_json_with_jq(input_path)
    

File: json_compression.js
const fs = require("fs");
const path = require("path");

function processSingleFile(filePath) {
    try {
        const content = fs.readFileSync(filePath, "utf8");
        const compressed = JSON.stringify(JSON.parse(content));

        const tempFile = path.join(
            path.dirname(filePath),
            `${path.basename(filePath, ".json")}_temp.json`
        );

        fs.writeFileSync(tempFile, compressed);
        fs.renameSync(tempFile, filePath);

        console.log(`JSON 文件 ${filePath} 压缩成功`);
        return 1;
    } catch (error) {
        console.error(`压缩失败 [${filePath}]: ${error.message}`);
        return 0;
    }
}

function compressJson(inputPath) {
    try {
        const stats = fs.statSync(inputPath);

        if (stats.isFile() && inputPath.toLowerCase().endsWith(".json")) {
            return processSingleFile(inputPath);
        } else if (stats.isDirectory()) {
            return processDirectory(inputPath);
        } else {
            console.error(`错误: ${inputPath} 不是有效的 JSON 文件或文件夹。`);
            return 0;
        }
    } catch (error) {
        console.error(`操作失败: ${error.message}`);
        return 0;
    }
}

function processDirectory(dirPath) {
    let successCount = 0;
    let failCount = 0;

    // 修改后的递归遍历函数
    function walk(currentDir) {
        const entries = fs.readdirSync(currentDir, { withFileTypes: true });
        for (const entry of entries) {
            const fullPath = path.join(currentDir, entry.name);
            if (entry.isDirectory()) {
                walk(fullPath); // 递归处理子目录
            } else if (
                fullPath.toLowerCase().endsWith(".json") &&
                !fullPath.includes("node_modules") && // 过滤依赖目录
                path.basename(fullPath) !== "package.json" // 过滤项目配置文件
            ) {
                try {
                    const result = processSingleFile(fullPath);
                    result ? successCount++ : failCount++;
                } catch (error) {
                    console.error(
                        `处理文件失败 [${fullPath}]: ${error.message}`
                    );
                    failCount++;
                }
            }
        }
    }

    walk(dirPath); // 初始调用

    console.log(`处理完成: 成功 ${successCount} 个, 失败 ${failCount} 个`);
    return successCount > 0 ? 1 : 0;
}

// 保持原有调用方式
if (require.main === module) {
    const inputPath = process.cwd();
    const result = compressJson(inputPath);
    module.exports = result;
} else {
    const inputPath = process.cwd();
    const result = compressJson(inputPath);
    module.exports = result;
}
